### PR TITLE
Phoenix - Tram H.

### PR DESCRIPTION
**Pull request by tramhn001**

### CODE REVIEW

File: adagrams/game.py
Comment (anselrognlie) [inline adagrams/game.py:1]: Try to commit more frequently on future projects. After finishing up each wave is a great time to commit your work so far.

File: adagrams/game.py [hunk 1]
```diff
@@ -1,11 +1,118 @@
...
+    'Z': 1
+}
+
 def draw_letters():
-    pass
+    letter_list = []
+    
+    # Create a list of letters, each letter appears according to their frequency
+    for letter, letter_freq in LETTER_POOL.items():
+        for i in range(letter_freq):
```
Comment (anselrognlie) [inline adagrams/game.py]: Notice that we don't use the variable `i` in the loop body. Python _does_ require that we provide a loop variable name here, but if we're not intending to actually use it, a common variable name is just the underscore character `_`. This tells the reader of our code that the only reason that variable is there is because Python requires it, not because we're doing anything with it.

This is most common in `for _ in range(some_value):` loops, where we just need to run the code in the block `some_value` times, without needing to know which iteration we're on while doing it.

File: adagrams/game.py [hunk 2]
```diff
@@ -1,11 +1,118 @@
...
+            letter_list.append(letter)
```
Comment (anselrognlie) [inline adagrams/game.py]: This code does a nice job of building up a list of all the available tiles. We could make this a little clearer by moving the logic to a helper function and giving it a good descriptive name, maybe `build_letter_list`.

File: adagrams/game.py [hunk 3]
```diff
@@ -1,11 +1,118 @@
...
+    
+
+    #Create an empty letter_bank list, each letter will be append to letter_bank according to their random index. 
+    #Using a while loop to control only ten letters with random indice will be added to letter_bank
+    #If an index is used, it will be added to the set of indiced_used, so it will not be chosen again
+    letter_bank = []
+    indices_used = set()
+    while len(letter_bank) < 10:
```
Comment (anselrognlie) [inline adagrams/game.py:58]: Since each loop might not result in adding an additional letter to the hand, we don't know exactly how many times the loop will run. Running until our hand has the desired number of tiles will make sure the loop runs as many times as necessary.

We could also consider giving a name to `10`, like `HAND_SIZE` so that it's clear what this number represents here. For a small project like this, we probably know what all the numbers represent, but for larger systems, having unnamed numbers showing up all over the place makes the code more difficult to understand.

File: adagrams/game.py [hunk 4]
```diff
@@ -1,11 +1,118 @@
...
+        index = random.randint(0, len(letter_list) - 1)
+        if index not in indices_used:
+            letter_bank.append(letter_list[index])
+            indices_used.add(index)
```
Comment (anselrognlie) [inline adagrams/game.py]: Notice that because we have to check whether the picked index has already been used, we aren't guaranteed to successfully pick a letter each time through the loop. Though unlikely, it's possible we could get "stuck" picking indices we've used already, never finishing picking a hand (or at least taking longer than expected). Even when there's a random process involved, we like to write logic that is guaranteed to make progress, rather than logic that could behave unpredictably.

Think about some approaches that we could use to ensure that we definitely pick a letter each time we iterate. If wepick an index that was already used, what could we do guarantee that we still pikc a letter? Or could we come up with a way so that any index we pick is guarantedd to be a valid, unused index?

File: adagrams/game.py [hunk 5]
```diff
@@ -1,11 +1,118 @@
...
 
 def uses_available_letters(word, letter_bank):
-    pass
+    letter_bank_dictionary = {}
+    # convert letter bank into a map between letter and its frequency
+    for letter in letter_bank:
+        if letter not in letter_bank_dictionary:
+            letter_bank_dictionary[letter]=1
+        else:
+            letter_bank_dictionary[letter]+=1
```
Comment (anselrognlie) [inline adagrams/game.py]: 👍 Great job building frequency maps for the bank. By building a new data structure that we use in our logic, we also avoid the possibility of accidentally modifying the hand passed in by the player.

Consider moving this logic to a helper function. If we had a function that could build a frequency map for a list, it would also be easy to use that for a string. If we had a frequency map for both the bank and the word, how could we use those to perform the validity check?

File: adagrams/game.py [hunk 6]
```diff
@@ -1,11 +1,118 @@
...
+    # convert input to upper case 
+    for letter in word.upper():
```
Comment (anselrognlie) [inline adagrams/game.py:71]: 👍 `upper` handles the case insensitive comparison. Because we're using English letters, upper case is fine. But for a more general notion of case-insensitive, take a look at `casefold`.

File: adagrams/game.py [hunk 7]
```diff
@@ -1,11 +1,118 @@
...
+        if letter not in letter_bank_dictionary or letter_bank_dictionary[letter] == 0:
+            return False
+        else:
+            letter_bank_dictionary[letter] -= 1
```
Comment (anselrognlie) [inline adagrams/game.py:75]: Since the `if` block terminates the immediate logic by returning, we can un-nest the main logic (it doesn't need to be under an `else`). `else` ensure that only one of the two blocks under an `if`/`else` is actually executed, by splitting the blocks into code that runs when the `if` conditional is true, and code that runs when it's false. Normally, if code that should only be run when the condition is false gets left unindented, it would get run regardless of the condition value
```py
        if some_condition:
            # only runs when condition is true
            some_logic()

        # runs whether the condition is true or false
        some_other_logic()
```

However, here, when the condition is true, we exit the function, meaning the only way we _could_ reach the code after the condition is if the condition had been false. If it had been true, we would have exited the function!

```py
        if letter not in letter_bank_dictionary or letter_bank_dictionary[letter] == 0:
            return False
            
        letter_bank_dictionary[letter] -= 1
```

This may be a little confusing, but it's a very common pattern (and allows us to indent less in python), so it's important to get used to this.

File: adagrams/game.py [hunk 8]
```diff
@@ -1,11 +1,118 @@
...
+    return True
 
 def score_word(word):
-    pass
+    total_scores = 0
+    for letter in word.upper():
+        scores = letter_values.get(letter, 0)
```
Comment (anselrognlie) [inline adagrams/game.py:81]: The `get()` call effectively ignores non-scored characters. We didn't have any tests around this, so that's a fair choice for you to make. Personally, if I use a function and somehow I ended up passing in unexpected data (where did these non-letter tiles come from?) I would rather know about it as soon as possible (like from getting a `KeyError` if we just use `letter_values[letter]`) rather than having it potentially be aproblem later and now I have less of an idea of what's going on.

Basically, I like to be noisy when something goes wrong.

File: adagrams/game.py [hunk 9]
```diff
@@ -1,11 +1,118 @@
...
+# create a dictionary of Letter and Value of points for each letter
+letter_values = {
+    'A': 1, 'E': 1, 'I': 1, 'O': 1, 'U': 1, 'L': 1, 'N': 1, 'R': 1, 'S': 1, 'T': 1,
+    'D': 2, 'G': 2,
+    'B': 3, 'C': 3, 'M': 3, 'P': 3,
+    'F': 4, 'H': 4, 'V': 4, 'W': 4, 'Y': 4,
+    'K': 5,
+    'J': 8, 'X': 8,
+    'Q': 10, 'Z': 10
+}
```
Comment (anselrognlie) [inline adagrams/game.py:96]: 👍 Nice job storing the scores so that we can look them up quickly using the letter.

It's fine that this is global (large data values can make a function harder to read), but I'd suggest naming it in all caps to indicate this is a CONSTANT value.

Also, consider ordering the letters alphabetically. The layout here is very close to how it was shown in the README, but as a person, if I wanted to quickly find one of the letters (maybe I need to update a score), it would be helpful if the code was arranged alphabetically. It would also help us double check that we've got all the letters accounted for. Keep in mind that the primary consumer of code is _not_ the computer, it's other coders. So laying out the data and logic in a way that's easy for others to follow is important.

File: adagrams/game.py [hunk 10]
```diff
@@ -1,11 +1,118 @@
...
+    return True
 
 def score_word(word):
-    pass
+    total_scores = 0
+    for letter in word.upper():
+        scores = letter_values.get(letter, 0)
+        total_scores += scores
+    if len(word) > 6:
+        total_scores += 8
```
Comment (anselrognlie) [inline adagrams/game.py:84]: 👍 Nice handling of adding the bonus.

We can also write this to make it more clear _why_ we're doing this. Consider giving names to the "magic numbers" (hard coded literal values that appear in code) that are used here. We could use `NO_BONUS_LEN` for `6`, and `LENGTH_BONUS` for `8`. We could even move this to a helper function called something like `add_bonus_points`.

File: adagrams/game.py [hunk 11]
```diff
@@ -1,11 +1,118 @@
...
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    best_word = ""
+    highest_score = 0
+    for word in word_list:
+        score = score_word(word)
+        
+        #Choose the word with highest scores
+        if score > highest_score:
```
Comment (anselrognlie) [inline adagrams/game.py:105]: 👍 A higher score word always replaces the best word so far.

File: adagrams/game.py [hunk 12]
```diff
@@ -1,11 +1,118 @@
...
+            best_word = word
+            highest_score = score
+        elif score == highest_score:
+            #if the word has 10 letters, it wins over the non-10-letter words.
+            if len(word) == 10 and len(best_word) != 10:
+                best_word = word
+            #if all the words are non-10-letter words, the word with shortest length will win over the longer ones
+            elif len(word) < len(best_word) and len(best_word) != 10:
+                best_word = word
```
Comment (anselrognlie) [inline adagrams/game.py:114]: Notice that both of these conditions check that the winning length isn't already 10 (another possible CONSTANT). A word can _only_ beat another tied word if the word we already found wasn't length 10 already. So we could move the `and len(best_word) != 10` for both of these up to the `elif score == highest_score` line, basically encoding the check as "if the new word is tied and the old word _can possibly_ be beaten".

But also, notice how complicated all of these checks are!

A different approach is to break up the parts of the logic into different chunks rather than trying to do it all in one loop. We can think of another approach like:
1. Find the max score (any winning word must have this score)
2. Make a list of all the words that have the max score
3. If we found only a single word, that's the winner
4. Otherwise, look at each word
5. If we find a 10 letter word, that's the winner (nothing would be able to beat it)
6. Otherwise, find the shortest word, that's the winner

The tie-breaking logic becomes a little easier thinking of the problem this way! Plus, each of those steps might be able to be written in their own helper function, making it easier to read the intention behind the logic, rather than needing to work backwards from the code to the intent. You do have comments to help with that, but comments can lie. 😈

File: adagrams/game.py [hunk 13]
```diff
@@ -1,11 +1,118 @@
...
+            #Python witll automatically return the first word in list, if there is a tie between words (same scores, same length)
+            #This is naturally handled since the first word is not replaced! 
+
+    return(best_word, highest_score)
```
Comment (anselrognlie) [inline adagrams/game.py:118]: Nit: `return` isn't a function (it's a keyword), so leave a space between `return` and the value being returned. Technically, the parens themselves are optional for a tuple, but a lot of coders are more comfortable using them. Just keep in mind that its the comma that makes the tuple.
```py
    return (best_word, highest_score)
```
or
```py
    return best_word, highest_score
```

File: main.py
Comment (anselrognlie) [inline main.py:1]: Watch out for commiting unintentional changes. If `git status` shows a file with changes that we don't want to include (such as with a following `git add .`), we can drop the changes with `git restore path/to/file`, so here `git restore main.py`.

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Slack, our next 1:1, or office hours are preferred (rather than replying to my comments in GitHub) as I don't have email notifications turned on. Nice job!

File: adagrams/game.py
Comment (tramhn001) [reply adagrams/game.py:1]: Yay, I will commit more frequently on next project :)

File: adagrams/game.py (hunk 2 above)
Comment (tramhn001) [reply adagrams/game.py]: Thank you for the suggestion. build_letter_list is a better name for it. I am so bad at "giving names" to the functions or test cases. That's one more thing that I should pay more attention to.

File: adagrams/game.py (hunk 3 above)
Comment (tramhn001) [reply adagrams/game.py:58]: Whoa!! It's totally new knowledge for me today, I have never thought of that before. Thank you so much for the suggestion. Unnamed numbers can be very confusing if someone else is reading the code.
